parsons/openfield/openfield.py

```diff
@@ -0,0 +1,536 @@
+    ):
+        # Make a general GET request
+
+        resp = self.conn.get(
+            self._base_endpoint(endpoint, entity_id),
+            params=params,
+        )
+        if resp.status_code >= 400:
+            raise Exception(self.parse_error(resp, exception_message))
+        return resp.json()
+
+    def _base_post(self, endpoint, data, exception_message=None):
+        # Make a general POST request
+
+        resp = self.conn.post(
+            self._base_endpoint(endpoint),
+            data=json.dumps(data),
+        )
+
+        if resp.status_code >= 400:
+            raise Exception(self.parse_error(resp, exception_message))
+
+        # Not all responses return a json
+        try:
+            return resp.json()
+
+        except ValueError:
+            return None
+
+    def _base_put(
+        self,
+        endpoint,
+        entity_id=None,
+        data=None,
+        params=None,
+        exception_message=None,
+    ):
+        # Make a general PUT request
+
+        endpoint = self._base_endpoint(endpoint, entity_id)
+
+        resp = self.conn.put(endpoint, data=json.dumps(data), params=params)
+
+        if resp.status_code >= 400:
+            raise Exception(self.parse_error(resp, exception_message))
+
+        # Not all responses return a json
+        try:
+            return resp.json()
+
+        except ValueError:
+            return None
+
+    def _base_patch(
+        self,
+        endpoint,
+        entity_id=None,
+        data=None,
+        params=None,
+        exception_message=None,
+    ):
+        # Make a general PATCH request
+
+        endpoint = self._base_endpoint(endpoint, entity_id)
+
+        resp = self.conn.patch(
+            endpoint,
+            data=json.dumps(data),
+            params=params,
+        )
+
+        if resp.status_code >= 400:
+            raise Exception(self.parse_error(resp, exception_message))
+
+        # Not all responses return a json
+        try:
+            return resp.json()
+
+        except ValueError:
+            return None
+
+    def _base_delete(
+        self,
+        endpoint,
+        entity_id=None,
+        data=None,
+        exception_message=None,
+    ):
+        # Make a general DELETE request
+
+        endpoint = self._base_endpoint(endpoint, entity_id)
+
+        resp = self.conn.delete(
+            endpoint,
+            data=json.dumps(data) if data else None,
+        )
+
+        if resp.status_code >= 400:
+            raise Exception(self.parse_error(resp, exception_message))
+
+        # Not all responses return a json
+        try:
+            return resp.json()
+
+        except ValueError:
+            return None
+
+    def parse_error(self, resp, exception_message):
+        """
+        Parse error responses from the API
+        """
+        message = f"Status {str(resp.status_code)}"
+        if exception_message:
+            message += exception_message
+
+        try:
+            json = resp.json()
```

Comment:
@sjwmoveon would like your advice on this: 

On 400 errors, some OpenField endpoints such as the bulk upsert people return a list of dicts that map to the indexes of the provided list, and will contain the columns containing bad data if there are any.

For example, a parsed response here could look like:
```[{}, {'phone1': ['The phone number entered is not valid.']}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {'phone1': ['The phone number entered is not valid.']}, {}, {}, {}]```

Right now, this just stringifies that response and adds it as a newline to the Exception message, under the status code.

In my end-consumer script which needs to use that list to identify the problem data and remove it before retrying the call, I end up needing to process the response like this:

```
if "400" in str(e):
  parsed_error_rows = ast.literal_eval(str(e).splitlines()[1].strip())
  
  for idx, row in enumerate(parsed_error_rows):
      for key in row:
          # do something with the bad data
```

Are you aware of a best practice that fits within the standards of how parsons returns errors that we could employ here to bubble both the status code and json in a way that's easier to parse for consumers?

Reply:
I think I figured out the optimal way to do this by extending `Exception` and appending relevant data from the response:
https://github.com/move-coop/parsons/pull/1044/files#diff-14899469f7551d91402e8a36a2d18b2ba509a00a8674092b7b7ab82eb7ef708bR11

This way, the consumer then looks like:
```
 except Exception as e:
            if e.status_code == 400:
                if e.json and isinstance(e.json, list):
                    for idx, row in enumerate(e.json):
                        for key in row:
                            # do something
```